src/rust/protocol/src/memcache/wire/request/parse.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use super::super::*;
+use crate::*;
+
+use core::slice::Windows;
+use std::convert::TryFrom;
+
+const MAX_BYTES: usize = usize::MAX / 2;
+
+impl Parse for MemcacheRequest {
+    fn parse(buffer: &[u8]) -> Result<ParseOk<Self>, ParseError> {
+        match parse_command(buffer)? {
+            MemcacheCommand::Get => parse_get(buffer),
+            MemcacheCommand::Gets => parse_gets(buffer),
+            MemcacheCommand::Set => parse_set(buffer, false),
+            MemcacheCommand::Add => parse_add(buffer),
+            MemcacheCommand::Replace => parse_replace(buffer),
+            MemcacheCommand::Cas => parse_set(buffer, true),
+            MemcacheCommand::Delete => parse_delete(buffer),
+        }
+    }
+}
+
+struct ParseState<'a> {
+    single_byte: Windows<'a, u8>,
+    double_byte: Windows<'a, u8>,
+}
+
+impl<'a> ParseState<'a> {
+    fn new(buffer: &'a [u8]) -> Self {
+        let single_byte = buffer.windows(1);
+        let double_byte = buffer.windows(2);
+        Self {
+            single_byte,
+            double_byte,
+        }
+    }
+
+    fn next_space(&mut self) -> Option<usize> {
+        self.single_byte.position(|w| w == b" ")
+    }
+
+    fn next_crlf(&mut self) -> Option<usize> {
+        self.double_byte.position(|w| w == CRLF.as_bytes())
+    }
+}
+
+fn parse_command(buffer: &[u8]) -> Result<MemcacheCommand, ParseError> {
+    let command;
+    {
+        let mut parse_state = ParseState::new(buffer);
+        if let Some(_line_end) = parse_state.next_crlf() {
+            if let Some(cmd_end) = parse_state.next_space() {
+                command = MemcacheCommand::try_from(&buffer[0..cmd_end])?;
+            } else {
+                return Err(ParseError::Invalid);
+            }
+        } else {
+            return Err(ParseError::Incomplete);
+        }
+    }
+    Ok(command)
+}
+
+#[allow(clippy::unnecessary_wraps)]
+fn parse_get(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
+    let mut parse_state = ParseState::new(buffer);
+
+    // this was already checked for when determining the command
+    let line_end = parse_state.next_crlf().unwrap();
+    let cmd_end = parse_state.next_space().unwrap();
+
+    let mut previous = cmd_end + 1;
+    let mut keys = Vec::new();
+
+    // command may have multiple keys, we need to loop until we hit
+    // a CRLF
+    loop {
+        if let Some(key_end) = parse_state.next_space() {
```

Comment:
I think you might need to absorb extra spaces between command and key, and between different keys to be safe.

Space is not allowed in keys for memcached ([ref](https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L48)) but current implementation will pick up the second whitespace as a key. e.g. `get  \r\n` will be considered legal by current impl with the key being a whitespace.

Reply:
I'm pretty sure the case you mention here does not work that way. There's an explicit test for `get \r\n` at https://github.com/twitter/pelikan/pull/312/files#diff-e41055767d02758e56bd95e779bbb9342f7b46f10c1d9412f3305e8bdb3f0b81R158

That said - `delete \r\n` is considered valid and returns a zero length key. I'm adding some basic validation of the response in the fuzzer - this will allow us to validate the parsing logic for test cases we haven't considered too.

Reply:
Trying to get away without absorbing extra spaces - I feel like it's going to complicate the DoS protections. I think the parser is a lot more bulletproof now. See if there are other validations we should add to the fuzzer code - it's really been helpful in exposing some odd corner cases I hadn't considered in standalone tests.